File: boj/TopologicalSorting/BOJ_2252.py
# 백준 2252 줄 세우기
# 21.03.16
import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
rd = sys.stdin.readline
INF = int(1e9)

n, m = map(int, input().split())
indegree = [0] * (n + 1)
edges = [[] for _ in range(n + 1)]

for _ in range(m):
    u, v = map(int, rd().split())
    edges[u].append(v)
    indegree[v] += 1

ans = []
q = []
for i in range(1, n + 1):
    if indegree[i] == 0:
        q.append(i)
        ans.append(i)

while q:
    now = q.pop(0)

    for i in edges[now]:
        indegree[i] -= 1
        if indegree[i] == 0:
            q.append(i)
            ans.append(i)

for i in ans:
    print(i, end=' ')

File: boj/TopologicalSorting/BOJ_2252.swift
// 백준 2252 줄 세우기
// 21.03.23

import Foundation

let nm = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = nm[0]
let m = nm[1]

var indegree = [Int](repeating: 0, count: n + 1)
var ans = [Int]()
var graph = [[Int]](repeating: [Int](), count: n + 1)

for _ in 0..<m {
  let arr = readLine()!.split(separator: " ").map { Int(String($0))! }
  graph[arr[0]].append(arr[1])
  indegree[arr[1]] += 1
}

var q = [Int]()

for i in 1...n {
  if indegree[i] == 0 {
    q.append(i)
    ans.append(i)
  }
}

while !q.isEmpty {
  let now = q.removeLast()
  
  for i in graph[now] {
    indegree[i] -= 1
    if indegree[i] == 0 {
      q.append(i)
      ans.append(i)
    }
  }
}

for i in ans { print(i, terminator: " ") }
